nucypher/blockchain/eth/sol/source/contracts/WorkLock.sol

```diff
@@ -109,74 +143,324 @@ contract WorkLock {
+
+        uint256 length = _biddersForRefund.length;
+        require(length > 0, "Must be at least one bidder for a refund");
+
+        uint256 minNumberOfBidders = tokenSupply.divCeil(maxAllowableLockedTokens);
+        if (bidders.length < minNumberOfBidders) {
+            shutdown();
+            return;
+        }
+
+        address previousBidder = _biddersForRefund[0];
+        uint256 minBid = workInfo[previousBidder].depositedETH;
+        uint256 maxBid = minBid;
+
+        // get minimum and maximum bids
+        for (uint256 i = 1; i < length; i++) {
+            address bidder = _biddersForRefund[i];
+            uint256 depositedETH = workInfo[bidder].depositedETH;
+            require(bidder > previousBidder && depositedETH > 0, "Addresses must be an array of unique bidders");
+            if (minBid > depositedETH) {
+                minBid = depositedETH;
+            } else if (maxBid < depositedETH) {
+                maxBid = depositedETH;
+            }
+            previousBidder = bidder;
+        }
+
+        uint256[] memory refunds = new uint256[](length);
+        // first step - align at a minimum bid
+        if (minBid != maxBid) {
+            for (uint256 i = 0; i < length; i++) {
+                address bidder = _biddersForRefund[i];
+                WorkInfo storage info = workInfo[bidder];
+                if (info.depositedETH > minBid) {
+                    refunds[i] = info.depositedETH - minBid;
+                    info.depositedETH = minBid;
+                    bonusETHSupply -= refunds[i];
+                }
+            }
+        }
+
+        require(ethToTokens(minBid) > maxAllowableLockedTokens,
+            "At least one of bidders has allowable bid");
+
+        // final bids adjustment (only for bonus part)
+        // (min_whale_bid * token_supply - max_stake * eth_supply) / (token_supply - max_stake * n_whales)
+        uint256 maxBonusTokens = maxAllowableLockedTokens - minAllowableLockedTokens;
+        uint256 minBonusETH = minBid - minAllowedBid;
+        uint256 bonusTokenSupply = tokenSupply - bidders.length * minAllowableLockedTokens;
+        uint256 refundETH = minBonusETH.mul(bonusTokenSupply)
+                                .sub(maxBonusTokens.mul(bonusETHSupply))
+                                .divCeil(bonusTokenSupply - maxBonusTokens.mul(length));
+        uint256 resultBid = minBid.sub(refundETH);
+        bonusETHSupply -= length * refundETH;
+        for (uint256 i = 0; i < length; i++) {
+            address bidder = _biddersForRefund[i];
+            WorkInfo storage info = workInfo[bidder];
+            refunds[i] += refundETH;
+            info.depositedETH = resultBid;
+        }
+
+        // reset verification
+        nextBidderToCheck = 0;
+
+        // transfer a refund
+        for (uint256 i = 0; i < length; i++) {
+            address payable bidder = _biddersForRefund[i];
+            bidder.sendValue(refunds[i]);
```

Comment:
There's no need to use the Withdrawal pattern here, right? My understanding is that it's not necessary since if a bidder that needs to be refunded is trying to do something fishy using a contract that doesn't accept ETH, what would happen is that it wouldn't be able to claim either. Just want to double-check.

Reply:
After reviewing this a second time, I think we should follow the withdrawal pattern here. We should do the ETH transfers on the claiming function, on a per-bidder basis.

Reply:
Good thoughts, I think still could be one irrational behaviour - make receive eth impossible and it will freeze worklock. so I think we need withdraw pattern for that, something very simple - mapping (address => refund)

Reply:
✔️